File: Arrays_&_Hashing/Longest_Consecutive_Sequence/3_Optimal.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> s(nums.begin(), nums.end());
        int best = 0;
        for (int x : s) {
            // only start counting if x-1 is not in the set
            if (!s.count(x - 1)) {
                int cur = 1;
                int y = x + 1;
                while (s.count(y)) {
                    ++cur;
                    ++y;
                }
                best = max(best, cur);
            }
        }
        return best;
    }
};

File: Arrays_&_Hashing/Longest_Consecutive_Sequence/3_Optimal.java
import java.util.*;

class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int num : nums) set.add(num);
        int best = 0;
        for (int x : set) {
            if (!set.contains(x - 1)) { // start of a sequence
                int cur = 1;
                int y = x + 1;
                while (set.contains(y)) {
                    cur++;
                    y++;
                }
                best = Math.max(best, cur);
            }
        }
        return best;
    }
}
